### PR TITLE
ISIS Kafka Listener Implementation

### PR DESCRIPTION
**Pull request by LamarMoore**
**NB. This PR Must be merged with the following third-party [PR](https://github.com/mantidproject/thirdparty-msvc2015/pull/15)!!**

This is the implementation of the a new Live Listener for Mantid which makes use of the [Kafka](https://kafka.apache.org/) and google [flatbuffers](https://google.github.io/flatbuffers/) technologies. This coincides with the development of a kafka based live streaming system which is under development at ISIS. This design will go hand-in-hand with the live stream set-up at the ESS. There were several levels to this development work which included:
 
 - Creating a [cluster](http://shizune:9000/clusters/isis_test_cluster/topics) which could act as a broker for the data stream.
- Creating a [producer](https://github.com/ScreamingUdder/isis_nexus_streamer_for_mantid) which streams large nexus files over the cluster which was used to test the live data stream.
- Laying the groundwork for the listener in mantid.
- Dependency configuration and functionality tests were also performed across all supported platforms along with ensuring packages were correctly built.

**To test:**

- Extensive Code Review.
- Ideally test that builds work on all platforms as well as package generation.

Fixes #xxxx.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by matthew-d-jones** [comment]
To test (must be at ISIS):
Set facility to `TEST_LIVE` and instrument to `ISIS_Kafka_Event`
Open the live data interface, `Load`->`Live Data` in the Workspaces toolbar.
Match these settings:
livedata_settings

Then hit `Run`.
After a few seconds you should have a workspace containing 14258850 events from the SANS instrument:
result_ws

**Comment by matthew-d-jones** [comment]
@martyngigg Could you give this PR a look when you have opportunity? I _think_ we are just missing rdkafka on build servers now.

**Comment by martyngigg** [comment]
@matthew-d-jones Sorry I haven't got to this yet.

It looks like a conflict has arisen though.

**Comment by matthew-d-jones** [comment]
@martyngigg No worries. It's not very urgent, I just want to get the Kafka listener in, in the hope that it'll reduce the amount of time I have to spend fixing merge conflicts when pulling upstream changes into the fork.

**Comment by martyngigg** [comment]
@matthew-d-jones Agreed.

I just have to coordinate getting the libraries installed on all of the builders first.

**Comment by martyngigg** [comment]
In fact, the Windows ones should take care of themselves as they were put in to the third party repository already so that build should go through.

**Comment by matthew-d-jones** [comment]
@martyngigg I realise that isn't a small task. Perhaps I should update the version of librdkafka first... we can talk about it at the meeting.

File: buildconfig/CMake/Bootstrap.cmake [hunk 1]
```diff
@@ -10,7 +10,7 @@ if( MSVC )
   include ( ExternalProject )
   set( EXTERNAL_ROOT ${PROJECT_SOURCE_DIR}/external )
   set( THIRD_PARTY_GIT_URL "https://github.com/mantidproject/thirdparty-msvc2015.git" )
-  set ( THIRD_PARTY_GIT_SHA1 4691e740fa2050f4fd0b44628eac445906bb2978 )
+  set ( THIRD_PARTY_GIT_SHA1 ac57236e560ef78a60cdec32503c26cf013cdf08 )
```
Comment (martyngigg) [inline buildconfig/CMake/Bootstrap.cmake]: @matthew-d-jones I've updated the Windows Kafka libraries to remove the LZ4 dependency. Can you update the SHA1 here to `09789e5ed0309f94b23eb4c8fa5e76ba07dc262b`.

Comment (martyngigg) [inline buildconfig/CMake/Bootstrap.cmake]: @matthew-d-jones  I managed to somehow screw up that last commit to the third party repo and some of the dlls stopped being tracked by git lfs. I think rebased and it managed to mess it up somehow. I've fixed that now but the SHA will need updating to `87a8fa24f066725494454c4ff10bcc21088b316a` now.

**Comment by martyngigg** [comment]
@matthew-d-jones Should we remove the in progress label?

**Comment by martyngigg** [comment]
I was just trying the test instructions and when i hit start it says it can't find `fewdatatest_events` on the broker. Has the topic name changed?

**Comment by matthew-d-jones** [comment]
Sorry, it seems I misremembered the name when I restreamed the data in after updating the brokers. It should be `fewtestdata`.

File: Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+ * @return A pointer to the data collected since the last call to this
+ * method
+ */
+API::Workspace_sptr KafkaEventStreamDecoder::extractData() {
+  if (m_exception) {
+    throw * m_exception;
+  }
+
+  m_extractWaiting = true;
+  cv.notify_one();
```
Comment (martyngigg) [inline Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp]: Nice use of condition variables. Does the static nature of the variable mean that we can then only have 1 listener running at a time?

(Admittedly I'm not sure if users would every have more 1 running...)

Comment (martyngigg) [inline Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp]: I feel a bit uneasy leaving these at file scope. Can we move them to member variables?

In examples for[ `condition_variable`](http://en.cppreference.com/w/cpp/thread/condition_variable) they tend to live at the same scope as the mutex.

**Review by martyngigg — CHANGES_REQUESTED** [review]
Other than the minor comments on the condition variable I think this looks really good.

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
I'm happy with this.

LGTM :shipit:

**Pull request merged**